Documentation (#10)
stlab/adobe-contract-checks: README.md

bugs early and prevent their damaging effects.

## Basic C++ Usage

This is a header-only library.  To use it from C++, simply put the
`include` directory of this repository in your `#include` path, and
`#include <adobe/contract_checks.hpp>`.

```c++
#include <adobe/contract_checks.hpp>
```

Every executable needs a single contract violation handler that
determines the program's behavior when a violation is detected.  A
good starting point is provided by a macro that you can expand in a
source file such as the one containing your `main` function.

```c++
ADOBE_DEFAULT_CONTRACT_VIOLATION_HANDLER() // no semicolon
```

There are three primary macros used to check contracts, each with two
forms: `ADOBE_PRECONDITION`, `ADOBE_POSTCONDITION`, and
`ADOBE_INVARIANT`. Each has one required argument and one optional
argument:

- `condition`: an expression convertible to `bool`; if `false`, the
  violation handler is invoked.
- `message`: an expression convertible to `const char*` pointing to a
  [null-terminated](https://en.cppreference.com/w/cpp/string/byte)
  message that is additionally passed to the violation handler. The

>   default value is the empty string, `""`.

Reviewer: Is `nullptr` allowed for the message?

Author: Does `nullptr` point to a null-terminated byte string?  I think not.

> In a macro I believe we can force a string literal by using "" message.
I don't understand what you're suggesting.

Author: @sean-parent I get it now.  But why would we want to do that?

Reviewer: By ensuring a string literal, you don't have to deal with null and lifetime if it escapes (on the down side, you can't build a message at runtime).

Author: It can only escape if you return or throw from your handler.  Are you sure it's worth it?  People can do any number of other bad things in their handlers too.

For example,

```c++
#include <adobe/contract_checks.hpp>
#include <climits>

// A half-open range of integers.
// - Invariant: start() <= end().
class int_range {
  // The lower bound; if the range is non-empty, its
  // least contained value.
  int _start;
  // The upper bound; if the range is non-empty, one
  // greater than its greatest contained value.
  int _end;

>   void check_invariant() { ADOBE_INVARIANT(start() <= end()); }

Reviewer: `check_invariant()` should be const. We should provide a bit more guidance about where to check class invariants.

Author: Good catch. There's a lovely suggestion feature that lowers the labor associated with making fixes like this.  Suggest you use it next time:

```suggestion
  void check_invariant() const { ADOBE_INVARIANT(start() <= end()); }
```

public:
  // An instance with the given bounds.
  // Precondition: end >= start
  int_range(int start, int end) : _start(start), _end(end) {
    ADOBE_PRECONDITION(end >= start, "invalid range bounds.");
    check_invariant();
  }

  // Returns the lower bound: if *this is non-empty, its
  // least contained value.
  int start() const { return _start; }

  // Returns the upper bound; if *this is non-empty, one
  // greater than its greatest contained value.
  int end() const { return _end; }

  // Increases the upper bound by 1.
  // Precondition: end() < INT_MAX
  void grow_rightward() {
    ADOBE_PRECONDITION(end() < INT_MAX);
    int old_end = end();
    _end += 1;
    ADOBE_POSTCONDITION(end() == old_end + 1);
    check_invariant();
  }

  // more methods...
};
```

## CMake Usage

  ```cmake
  include(FetchContent)
  if(PROJECT_IS_TOP_LEVEL)
    FetchContent_Declare(
      adobe-contract-checks
      GIT_REPOSITORY https://github.com/stlab/adobe-contract-checks.git
      GIT_TAG        <this library's release version>
    )
    FetchContent_MakeAvailable(adobe-contract-checks)
  endif()
  find_package(adobe-contract-checks)

  add_library(my-library my-library.cpp)

>   target_link_libraries(my-library PRIVATE adobe-contract-checks)

Reviewer: I'm assuming `target_link_libraries` is adding it to the include paths - is there any way to add it to the include paths for all targets instead of repeating these?

Author: Yes.  And no, not a principled way.  This is the right way to do it.

Author: Even if only some of those targets depend on the library supplying those includes?  I don't think that's principled at all.

Author: I'm calling this resolved.  Reopen if you think there's something more to say.

   // by "<year>-<month>-<day>" (interpreted in ISO standard date
   // format)
   day_of_the_week day(int year, int month, int day) {
     ADOBE_PRECONDITION(is_valid_date(year, month, day));
     // implementation starts here.
   }

   // Returns the day of the week corresponding to `d`.
   day_of_the_week day(date d) {
     // implementation starts here.
   }
   ```

   The second function above benefits by accepting a `date` type whose
   invariant ensures its validity.

4. The conditions in your checks should not have side-effects that
   change program behavior, because checks are sometimes turned off by
   configuration.

2. Group all precondition checks immediately after a function's
   opening brace, and don't allow any code to sneak in before them.

3. Group all postcondition checks just before your function returns.
   (That may mean temporarily storing a return value in a local
   variable so it can be tested.)

4. Give your `struct` or `class` a `void check_invariant() const`
   method containing `ADOBE_INVARIANT` invocations.  Invoke it just
   before returning from each public mutating friend or member

>    function.

Reviewer: > and before passing `this` as an argument to an external component from a mutating operation.

Author: using the suggestion feature would be appreciated here too.

>    method containing `ADOBE_INVARIANT` invocations.  Invoke it just
>    before returning from each public mutating friend or member
>    function.

Author: ```suggestion
   method containing `ADOBE_INVARIANT` invocations.  Invoke it from each 
   public mutating friend or member function, just before returning, and just before passing 
   access to `*this` to any component outside the class.
```

       = std::set_terminate(emergency_shutdown);

     [[noreturn]] void emergency_shutdown() noexcept
     {
       // emergency shutdown measures here.

       if (previous_terminate_handler != nullptr)
         { previous_terminate_handler(); }
       std::abort();
     }

     [[noreturn]] void ::adobe::contract_violated(
       const char *const condition,
       ::adobe::contract_violation::kind_t kind,
       const char *const file,
       std::uint32_t const line,
       const char *const message) noexcept
     {
       // whatever you want here.
       std::terminate();
     }
     ```

   That way, other reasons for unexpected termination, such as uncaught
   exceptions, will still cause emergency shutdown.

3. If your custom contract violation handler needs to print a
  description of the failure, use [Gnu standard error
  format](https://www.gnu.org/prep/standards/html_node/Errors.html#Errors),
  which  will be automatically understood by many tools.  The
  following expression will print such a report to the standard error
  stream:

      ```c++
      adobe::contract_violation(
        condition, kind, file, line, message).print_report();
      ```

> 5. Don't disable critical checks in shipping code unless an measurable

Reviewer: "a measurable"

> 5. Don't disable critical checks in shipping code unless an measurable

Author: ```suggestion
5. Don't disable critical checks in shipping code unless a measurable
```

  Maintenance is more error-prone and quality suffers.

Note that in an unsafe language like C++, a seemingly recoverable
condition like the discovery of a negative index can easily be the
result of undefined behavior that also scrambled memory or causes
“impossible” execution.

## Development

The usual procedures for development with cmake apply.  One typical
set of commands might be:

```sh
cmake -Wno-dev -S . -B ../build -GNinja        # configure
cmake --build ../build                         # build/rebuild after changes
ctest --output-on-failure --test-dir ../build  # test
```

## Reference

### Configuration

The behavior of this library is configured by preprocessor symbols.

Instead of passing many `-D` options in your compiler command-line,
you can put the definition of these symbols in a header file and
simply define `ADOBE_CONTRACT_CHECKS_CONFIGURATION` as the `#include`
argument to that file:

```sh
c++ -D \
    -I .\
    ADOBE_CONTRACT_CHECKS_CONFIGURATION="<myproject/adobe_contract_checks_config.hpp>" \
    myproject_source.cpp -o myproject_executable
```

If you are using this project via CMake, defining the same symbol in

> CMake, on the command line or in your will cause all clients of this

Reviewer: "or in your" what?

Author: ```suggestion
your `CMakeLists.txt` or on the command line will cause all clients of this
```

library in your build to use that configuration file.

This library can only have one configuration in an executable, so the
privilege of configuring it belongs to the executable.  In CMake,

```cmake
if(PROJECT_IS_TOP_LEVEL)
  set(ADOBE_CONTRACT_CHECKS_CONFIGURATION "<myproject/adobe_contract_checks_config.hpp>")
endif()
```

#### Preprocessor configuration symbols

##### Contract handler definition

- `ADOBE_CONTRACT_VIOLATED_INLINE_BODY`: if you want the contract
  violation handler to be defined inline, make this symbol expand to
  its body.  When the body is lighter-weight than a call to the
  handler would be, an inline handler can limit the code generated at
  the use site. For example,

  ```c++
  #define ADOBE_CONTRACT_VIOLATED_INLINE_BODY { ADOBE_BUILTIN_TRAP(); }
  ```

  You might use this option in release builds when a minimal handler
  is required.  **Note:** Defining a more complex handler inline
  usually will increase binary sizes and may hurt performance.

> - `ADOBE_CONTRACT_VIOLATED_RETURNS`: define this symbol if your

Reviewer: Do we _ever_ want this option? I could see the `INLINE_BODY` returning an error from the enclosing function, but that would require a mandate on how errors are returned from all functions using the library. Do we have an example?

Author: I'm happy to remove this option, but I can imagine the same people that want an exception wanting to log the problem and continue, which after all is slightly better than what happens when you compile assertions out.  Your call.

Reviewer: I'm good with dropping it.

Author: Done.
